django/db/models/sql/query.py

```diff
@@ -289,8 +289,9 @@ class Query(BaseExpression):
 
     explain_info = None
 
-    def __init__(self, model, alias_cols=True):
+    def __init__(self, model, alias_cols=True, queryset=None):
```

Comment:
I think I would prefer not to pass in the queryset and then we can be less worried about evaluating it, maybe we can take a look at the db alias?
Something like:
```diff
--- a/django/db/models/sql/query.py
+++ b/django/db/models/sql/query.py
@@ -288,10 +288,10 @@ class Query(BaseExpression):
     deferred_loading = (frozenset(), True)

     explain_info = None
+    using = None

-    def __init__(self, model, alias_cols=True, queryset=None):
+    def __init__(self, model, alias_cols=True):
         self.model = model
-        self.queryset = queryset
         self.alias_refcount = {}
         # alias_map is the most important data structure regarding joins.
         # It's used for recording which joins exist in the query and what
@@ -318,19 +318,9 @@ class Query(BaseExpression):

         self._filtered_relations = {}

-    # Remove queryset from pickle result.
-    # Refer https://code.djangoproject.com/ticket/27159
-    def __getstate__(self):
-        state = self.__dict__.copy()
-        state["queryset"] = None
-        return state
-
     @property
     def db(self):
-        # Must not use ``if self.queryset``, that will trigger evaluation of queryset.
-        if self.queryset is not None:
-            return self.queryset.db
-        return DEFAULT_DB_ALIAS
+        return self.using or DEFAULT_DB_ALIAS
```
and then we set `query.using` 🤔 maybe we call it `_db` or `db` (naming is hard)

Reply:
Thank you for your suggestion, in the last commit, I have pass QuerySet.db instead of QuerySet itself to Query.